core/src/font.rs

```diff
@@ -1210,9 +1252,98 @@ impl Default for TextRenderSettings {
     }
 }
 
+/// Font set contains a set of fonts used to render text.
+///
+/// It always contains at least one font—the main font. It may also contain
+/// fallback fonts, which will be used in case glyphs are missing from the main
+/// font. Fallback fonts are always used in order.
+#[derive(Debug, Clone, Collect, Copy)]
+#[collect(no_drop)]
+pub struct FontSet<'gc>(Gc<'gc, FontSetData<'gc>>);
+
+#[derive(Debug, Collect)]
+#[collect(no_drop)]
+struct FontSetData<'gc> {
+    main_font: Font<'gc>,
+    fallback_fonts: Vec<Font<'gc>>,
+}
+
+impl<'gc> FontSet<'gc> {
+    /// Creates a font set from a sorted list of fonts.
+    ///
+    /// The first font is the main font, the rest are fallbacks.
+    ///
+    /// Returns None when the list is empty.
+    pub fn from_fonts(mc: &Mutation<'gc>, fonts: Vec<Font<'gc>>) -> Option<Self> {
+        let (&main_font, fallback_fonts) = fonts.split_first()?;
```

Comment:
Nit: this could be simplified to
```rs
let main_font = fonts.remove(0);
Some(Self(Gc::new(
    mc,
    FontSetData {
        main_font,
        fallback_fonts: fonts,
    },
)))
```